First retried task listener after `EXTRACT_VALUE_ERROR` incident loses original command context

### <!-- Component -->

<!-- Zeebe- -->

### <!-- Affected version -->

<!-- -8.8 -->

### <!-- Severity -->

<!-- Low- -->

### <!-- Likelihood -->

<!-- Low_ -->

### Description

When a user task operation (e.g., `UPDATE`, `ASSIGN`, or `CLAIM`) triggers task listeners and one of them fails with an `EXTRACT_VALUE_ERROR`, the first listener job retried after resolving the incident does not see the data changed by the original command.

This leads to inconsistent behavior:
- The first retried listener receives headers with values reflecting the user task state before the current command was received.
- The second and subsequent listeners (if defined), receive the correct set of headers as were changed by the original command.

### Steps to reproduce

1. Create a process with a user task that has two `updating` task listeners.
2. One listener should have a valid `type`, another should use an expression that evaluates to a missing variable (`typeExpression("missing_var_name")`).
3. Perform a user task update with multiple attributes changed (`priority`, `candidateUsers`, `dueDate`, `action`).
4. An incident is created due to the missing variable in one listener's expression.
5. Resolve the incident by supplying the missing variable, listener jobs should be re-created and re-tryed.

**🧪 Test to reproduce:**
<details>
<summary>Click here to expand test</summary>

```java
  @Test
  public void retriedTaskListenerShouldHaveAccessToDataProvidedByOriginalCommand() {
    // given: process instance is created with a UT having `updating` listeners
    final long processInstanceKey =
        helper.createProcessInstance(
            helper.createProcessWithZeebeUserTask(
                task ->
                    task.zeebeTaskListener(l -> l.updating().type(listenerType))
                        .zeebeTaskListener(l -> l.updating().typeExpression("missing_var_name"))));

    // when:
    ENGINE
        .userTask()
        .ofInstance(processInstanceKey)
        .withCandidateUsers("bob", "alice")
        .withDueDate("new_due_date")
        .withPriority(88)
        .withAction("custom_update")
        .update();

    // then: assert expected header for the first listener job and complete it
    final Consumer<JobRecordValue> jobHeadersAssertion =
        job ->
            assertThat(job.getCustomHeaders())
                .contains(
                    entry(Protocol.USER_TASK_CANDIDATE_USERS_HEADER_NAME, "[\"bob\",\"alice\"]"),
                    entry(Protocol.USER_TASK_DUE_DATE_HEADER_NAME, "new_due_date"),
                    entry(Protocol.USER_TASK_PRIORITY_HEADER_NAME, "88"),
                    entry(
                        Protocol.USER_TASK_CHANGED_ATTRIBUTES_HEADER_NAME,
                        "[\"candidateUsersList\",\"dueDate\",\"priority\"]"),
                    entry(Protocol.USER_TASK_ACTION_HEADER_NAME, "custom_update"));
    helper.assertActivatedJob(processInstanceKey, listenerType, jobHeadersAssertion);
    ENGINE.job().ofInstance(processInstanceKey).withType(listenerType).complete();

    // then: expect an incident due to missing `assign_listener_var_name` variable
    final var incident =
        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
            .withProcessInstanceKey(processInstanceKey)
            .getFirst();

    Assertions.assertThat(incident.getValue())
        .hasProcessInstanceKey(processInstanceKey)
        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
        .hasErrorMessage(
            """
                Expected result of the expression 'missing_var_name' to be 'STRING', but was 'NULL'. \
                The evaluation reported the following warnings:
                [NO_VARIABLE_FOUND] No variable found with name 'missing_var_name'""");

    // when: fix the missing variable and resolve the incident
    ENGINE
        .variables()
        .ofScope(processInstanceKey)
        .withDocument(Map.of("missing_var_name", listenerType + "_2"))
        .update();
    ENGINE.incident().ofInstance(processInstanceKey).withKey(incident.getKey()).resolve();

    // then: assert and complete the retried task listener job
    helper.assertActivatedJob(processInstanceKey, listenerType, jobHeadersAssertion); // <-  ❌ fails here
    helper.completeRecreatedJobWithType(processInstanceKey, listenerType);

    // assert headers for the next job, they should be same as for the fist one
    helper.assertActivatedJob(processInstanceKey, listenerType + "_2", jobHeadersAssertion);
  }
```

fails with the following error:
```shell
java.lang.AssertionError: 
Expecting map:
  {"io.camunda.zeebe:priority"="50", "io.camunda.zeebe:userTaskKey"="2251799813685256"}
to contain entries:
  ["io.camunda.zeebe:candidateUsers"="["bob","alice"]",
    "io.camunda.zeebe:dueDate"="new_due_date",
    "io.camunda.zeebe:priority"="88",
    "io.camunda.zeebe:changedAttributes"="["candidateUsersList","dueDate","priority"]",
    "io.camunda.zeebe:action"="custom_update"]
but could not find the following map entries:
  ["io.camunda.zeebe:candidateUsers"="["bob","alice"]",
    "io.camunda.zeebe:dueDate"="new_due_date",
    "io.camunda.zeebe:changedAttributes"="["candidateUsersList","dueDate","priority"]",
    "io.camunda.zeebe:action"="custom_update"]
and the following map entries had different values:
  ["io.camunda.zeebe:priority"="50" (expected: "88")]
```
</details>

### Current behavior

The retried first listener receives headers with values reflecting the user task state before the current command was received.

### Expected behavior

The first retried listener should receive the same custom headers with values (reflecting the received user task command) as it did before the failure.

### Environment

SaaS & SM

### Version

- Zeebe: [8.8.0-SNAPSHOT]


### Rootcause

This issue occurs because after retrying the failed command, we skip the `UserTaskCommandProcessor#onCommand` method. As a result, we do not extract values from the command again. Instead, we use the `UserTask` state as it was before the command was processed to create the task listener job.

### Solution ideas

We can provide the `UserTask` from the intermediate state that was already built during the initial command processing (before the transition was interrupted by the incident). This intermediate state contains all the updated attributes, and by using it to re-create a listener job would ensure that it will receive the most recent headers.

### Dev -> QA handover

- Resources:
- Versions to validate: 8.8
- Release version (in which version this feature will be released): 8.8


### Links
- Related stretch goal (potentially solves this issue by skipping re-execution):
  - #24107

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobBehavior.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobBehavior.java
@@ -491,7 +491,7 @@ public final class BpmnJobBehavior {
     if (changedAttributes != null && !changedAttributes.isEmpty()) {
       headers.put(
           Protocol.USER_TASK_CHANGED_ATTRIBUTES_HEADER_NAME,
-          ExpressionTransformer.asListLiteral(changedAttributes));
+          ExpressionTransformer.asListLiteral(changedAttributes.stream().sorted().toList()));
     }
 
     if (userTaskRecord.getPriority() > 0) {

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/IncidentResolveProcessor.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/IncidentResolveProcessor.java
@@ -214,7 +214,7 @@ public final class IncidentResolveProcessor implements TypedRecordProcessor<Inci
             intent -> {
               final var userTaskRecord = new UserTaskRecord();
               userTaskRecord.wrap(intermediateState.getRecord());
-              return new IncidentRecordWrapper<>(userTaskKey, intent, userTaskRecord);
+              return new RetryTypedRecord<>(userTaskKey, intent, userTaskRecord);
             });
   }
 
@@ -226,7 +226,7 @@ public final class IncidentResolveProcessor implements TypedRecordProcessor<Inci
             intent -> {
               final var record = new ProcessInstanceRecord();
               record.wrap(elementInstance.getValue());
-              return new IncidentRecordWrapper<>(elementInstance.getKey(), intent, record);
+              return new RetryTypedRecord<>(elementInstance.getKey(), intent, record);
             });
   }
 

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/RetryTypedRecord.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/RetryTypedRecord.java
@@ -18,18 +18,24 @@ import io.camunda.zeebe.stream.api.records.TypedRecord;
 import java.util.Map;
 
 /**
- * A wrapper class for records that facilitates retrying commands after an incident is resolved.
+ * A minimal implementation of {@link TypedRecord} used to reprocess a command, typically after an
+ * incident has been resolved.
  *
  * <p>This class is designed to handle different types of records and their corresponding intents,
  * allowing seamless recovery by re-processing the failed command.
+ *
+ * <p>Consumers can use the type of this class to identify and handle retried commands differently,
+ * if special logic is needed during reprocessing.
+ *
+ * @param <T> the type of the record value being retried
  */
-final class IncidentRecordWrapper<T extends UnifiedRecordValue> implements TypedRecord<T> {
+public final class RetryTypedRecord<T extends UnifiedRecordValue> implements TypedRecord<T> {
 
   private final long key;
   private final Intent intent;
   private final T record;
 
-  IncidentRecordWrapper(final long key, final Intent intent, final T record) {
+  RetryTypedRecord(final long key, final Intent intent, final T record) {
     this.key = key;
     this.intent = intent;
     this.record = record;
@@ -67,7 +73,7 @@ final class IncidentRecordWrapper<T extends UnifiedRecordValue> implements Typed
 
   @Override
   public RecordType getRecordType() {
-    return null;
+    return RecordType.COMMAND;
   }
 
   @Override

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
@@ -168,31 +168,29 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
     final var commandProcessor = commandProcessors.getCommandProcessor(intent);
 
     if (isRetriedCommand(command)) {
-      // Skip validation for retried command, as it was already validated
-      // during the original execution
-      handleCommandProcessing(commandProcessor, command, command.getValue(), intent);
+      // Skip `validateCommand` and `onCommand` invocations for retried command,
+      // as it was already validated during the original execution
+      finalizeCommandOrCreateTaskListenerJob(commandProcessor, command, command.getValue(), intent);
     } else {
       commandProcessor
           .validateCommand(command)
+          // Create a modifiable copy of the persisted user task record, as `onCommand` may
+          // apply the changed attributes from the command on top of persisted instance.
+          .map(UserTaskRecord::copy)
+          .thenDo(persistedRecord -> commandProcessor.onCommand(command, persistedRecord))
           .ifRightOrLeft(
               persistedRecord ->
-                  handleCommandProcessing(
-                      commandProcessor, command, persistedRecord.copy(), intent),
+                  finalizeCommandOrCreateTaskListenerJob(
+                      commandProcessor, command, persistedRecord, intent),
               rejection -> handleCommandRejection(command, rejection));
     }
   }
 
-  private void handleCommandProcessing(
+  private void finalizeCommandOrCreateTaskListenerJob(
       final UserTaskCommandProcessor processor,
       final TypedRecord<UserTaskRecord> command,
       final UserTaskRecord persistedRecord,
       final UserTaskIntent intent) {
-
-    // For retried commands (reconstructed after an incident), we skip `onCommand()`
-    // since it was already executed during the original command handling before the failure.
-    if (!isRetriedCommand(command)) {
-      processor.onCommand(command, persistedRecord);
-    }
 
     final var userTaskElement = getUserTaskElement(persistedRecord);
     final var eventType = mapIntentToEventType(intent);

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
@@ -220,16 +220,6 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
     }
   }
 
-  /**
-   * Determines whether the given {@link TypedRecord} is a retried command reconstructed for
-   * reprocessing after an incident was resolved.
-   *
-   * <p>Used to skip redundant processing steps that were already performed during the original
-   * command execution.
-   *
-   * @param command the typed record to check
-   * @return {@code true} if the record represents a retried command
-   */
   private boolean isRetriedCommand(final TypedRecord<UserTaskRecord> command) {
     return command instanceof RetryTypedRecord<UserTaskRecord>;
   }

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
@@ -16,6 +16,7 @@ import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnJobBehavior;
 import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableUserTask;
 import io.camunda.zeebe.engine.processing.deployment.model.element.TaskListener;
 import io.camunda.zeebe.engine.processing.identity.AuthorizationCheckBehavior;
+import io.camunda.zeebe.engine.processing.incident.IncidentRecordWrapper;
 import io.camunda.zeebe.engine.processing.streamprocessor.TypedRecordProcessor;
 import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
 import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedRejectionWriter;
@@ -165,12 +166,20 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
   private void processOperationCommand(
       final TypedRecord<UserTaskRecord> command, final UserTaskIntent intent) {
     final var commandProcessor = commandProcessors.getCommandProcessor(intent);
-    commandProcessor
-        .validateCommand(command)
-        .ifRightOrLeft(
-            persistedRecord ->
-                handleCommandProcessing(commandProcessor, command, persistedRecord.copy(), intent),
-            rejection -> handleCommandRejection(command, rejection));
+
+    if (isRetriedCommand(command)) {
+      // Skip validation for retried command, as it was already validated
+      // during the original execution
+      handleCommandProcessing(commandProcessor, command, command.getValue(), intent);
+    } else {
+      commandProcessor
+          .validateCommand(command)
+          .ifRightOrLeft(
+              persistedRecord ->
+                  handleCommandProcessing(
+                      commandProcessor, command, persistedRecord.copy(), intent),
+              rejection -> handleCommandRejection(command, rejection));
+    }
   }
 
   private void handleCommandProcessing(
@@ -179,11 +188,9 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
       final UserTaskRecord persistedRecord,
       final UserTaskIntent intent) {
 
-    // If a user-triggered command (ASSIGN, CLAIM, UPDATE, COMPLETE) lacks request metadata,
-    // it indicates the command was retried by the engine after resolving an incident caused
-    // by a task listener property expression evaluation failure. In this case, the `onCommand`
-    // method was already processed during the initial execution, so we can safely skip it now.
-    if (command.hasRequestMetadata()) {
+    // For retried commands (reconstructed after an incident), we skip `onCommand()`
+    // since it was already executed during the original command handling before the failure.
+    if (!isRetriedCommand(command)) {
       processor.onCommand(command, persistedRecord);
     }
 
@@ -215,7 +222,21 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
     }
   }
 
-  void storeUserTaskRecordRequestMetadata(final TypedRecord<UserTaskRecord> command) {
+  /**
+   * Determines whether the given {@link TypedRecord} is a retried command reconstructed for
+   * reprocessing after an incident was resolved.
+   *
+   * <p>Used to skip redundant processing steps that were already performed during the original
+   * command execution.
+   *
+   * @param command the typed record to check
+   * @return {@code true} if the record represents a retried command
+   */
+  private boolean isRetriedCommand(final TypedRecord<UserTaskRecord> command) {
+    return command instanceof IncidentRecordWrapper<UserTaskRecord>;
+  }
+
+  private void storeUserTaskRecordRequestMetadata(final TypedRecord<UserTaskRecord> command) {
     if (!command.hasRequestMetadata()) {
       return;
     }

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java
@@ -16,7 +16,7 @@ import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnJobBehavior;
 import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableUserTask;
 import io.camunda.zeebe.engine.processing.deployment.model.element.TaskListener;
 import io.camunda.zeebe.engine.processing.identity.AuthorizationCheckBehavior;
-import io.camunda.zeebe.engine.processing.incident.IncidentRecordWrapper;
+import io.camunda.zeebe.engine.processing.incident.RetryTypedRecord;
 import io.camunda.zeebe.engine.processing.streamprocessor.TypedRecordProcessor;
 import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
 import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedRejectionWriter;
@@ -233,7 +233,7 @@ public class UserTaskProcessor implements TypedRecordProcessor<UserTaskRecord> {
    * @return {@code true} if the record represents a retried command
    */
   private boolean isRetriedCommand(final TypedRecord<UserTaskRecord> command) {
-    return command instanceof IncidentRecordWrapper<UserTaskRecord>;
+    return command instanceof RetryTypedRecord<UserTaskRecord>;
   }
 
   private void storeUserTaskRecordRequestMetadata(final TypedRecord<UserTaskRecord> command) {

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/processors/UserTaskCommandPreconditionChecker.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/processors/UserTaskCommandPreconditionChecker.java
@@ -62,19 +62,14 @@ public class UserTaskCommandPreconditionChecker {
 
   protected Either<Rejection, UserTaskRecord> check(final TypedRecord<UserTaskRecord> command) {
     final long userTaskKey = command.getKey();
-    final UserTaskRecord persistedRecord = fetchUserTaskRecord(command);
+    final var persistedRecord =
+        userTaskState.getUserTask(userTaskKey, authCheckBehavior.getAuthorizedTenantIds(command));
 
     if (persistedRecord == null) {
       return Either.left(
           new Rejection(
               RejectionType.NOT_FOUND,
               String.format(NO_USER_TASK_FOUND_MESSAGE, intent, userTaskKey)));
-    }
-
-    // Skip remaining checks for Zeebe-retried commands (missing request metadata),
-    // as they were already performed during the initial command processing.
-    if (!command.hasRequestMetadata()) {
-      return Either.right(persistedRecord);
     }
 
     final var authRequest =
@@ -101,13 +96,5 @@ public class UserTaskCommandPreconditionChecker {
         .map(checks -> checks.apply(command, persistedRecord))
         .filter(Either::isLeft)
         .orElse(Either.right(persistedRecord));
-  }
-
-  private UserTaskRecord fetchUserTaskRecord(final TypedRecord<UserTaskRecord> command) {
-    final long userTaskKey = command.getKey();
-    return command.hasRequestMetadata()
-        ? userTaskState.getUserTask(userTaskKey, authCheckBehavior.getAuthorizedTenantIds(command))
-        // For Zeebe-retried commands (missing metadata), fetch the user task without tenant checks.
-        : userTaskState.getUserTask(userTaskKey);
   }
 }

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerIncidentsTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerIncidentsTest.java
@@ -565,9 +565,9 @@ public class TaskListenerIncidentsTest {
             ENGINE
                 .userTask()
                 .ofInstance(pik)
-                .withCandidateUsers("bob", "alice")
                 .withDueDate("new_due_date")
                 .withPriority(88)
+                .withCandidateUsers("bob", "alice")
                 .withAction("custom_update")
                 .update(),
         Map.of(

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerIncidentsTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerIncidentsTest.java
@@ -15,6 +15,7 @@ import io.camunda.zeebe.engine.util.EngineRule;
 import io.camunda.zeebe.model.bpmn.Bpmn;
 import io.camunda.zeebe.model.bpmn.builder.UserTaskBuilder;
 import io.camunda.zeebe.model.bpmn.instance.zeebe.ZeebeTaskListenerEventType;
+import io.camunda.zeebe.protocol.Protocol;
 import io.camunda.zeebe.protocol.impl.record.value.usertask.UserTaskRecord;
 import io.camunda.zeebe.protocol.record.Assertions;
 import io.camunda.zeebe.protocol.record.Record;
@@ -25,6 +26,7 @@ import io.camunda.zeebe.protocol.record.intent.UserTaskIntent;
 import io.camunda.zeebe.protocol.record.intent.VariableIntent;
 import io.camunda.zeebe.protocol.record.value.ErrorType;
 import io.camunda.zeebe.protocol.record.value.JobListenerEventType;
+import io.camunda.zeebe.protocol.record.value.JobRecordValue;
 import io.camunda.zeebe.protocol.record.value.UserTaskRecordValue;
 import io.camunda.zeebe.test.util.record.RecordingExporter;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
@@ -40,6 +42,7 @@ import org.junit.rules.TestWatcher;
 
 /** Tests verifying that a task listener may raise resolvable incidents. */
 public class TaskListenerIncidentsTest {
+
   @ClassRule public static final EngineRule ENGINE = EngineRule.singlePartition();
   @Rule public final TestWatcher watcher = new RecordingExporterTestWatcher();
   private final TaskListenerTestHelper helper = new TaskListenerTestHelper(ENGINE);
@@ -444,7 +447,6 @@ public class TaskListenerIncidentsTest {
 
   @Test
   public void shouldRetryCancelingListenerJobAfterExtractValueErrorDuringUserTaskInterruption() {
-
     // given
     final long processInstanceKey =
         helper.createProcessInstance(
@@ -513,5 +515,132 @@ public class TaskListenerIncidentsTest {
         processInstanceKey,
         UserTaskIntent.CANCELED,
         userTask -> assertThat(userTask.getAction()).isEmpty());
+  }
+
+  @Test
+  public void retriedAssigningListenerJobShouldHaveCorrectHeadersFromAssignCommand() {
+    testRetriedListenerJobHasExpectedHeaders(
+        ZeebeTaskListenerEventType.assigning,
+        pik ->
+            ENGINE
+                .userTask()
+                .ofInstance(pik)
+                .withAssignee("bryan")
+                .withAction("custom_assignment")
+                .assign(),
+        Map.of(
+            Protocol.USER_TASK_ASSIGNEE_HEADER_NAME,
+            "bryan",
+            Protocol.USER_TASK_CHANGED_ATTRIBUTES_HEADER_NAME,
+            "[\"assignee\"]",
+            Protocol.USER_TASK_ACTION_HEADER_NAME,
+            "custom_assignment"));
+  }
+
+  @Test
+  public void retriedAssigningListenerJobShouldHaveCorrectHeadersFromClaimCommand() {
+    testRetriedListenerJobHasExpectedHeaders(
+        ZeebeTaskListenerEventType.assigning,
+        pik ->
+            ENGINE
+                .userTask()
+                .ofInstance(pik)
+                .withAssignee("jesse")
+                .withAction("custom_claiming")
+                .claim(),
+        Map.of(
+            Protocol.USER_TASK_ASSIGNEE_HEADER_NAME,
+            "jesse",
+            Protocol.USER_TASK_CHANGED_ATTRIBUTES_HEADER_NAME,
+            "[\"assignee\"]",
+            Protocol.USER_TASK_ACTION_HEADER_NAME,
+            "custom_claiming"));
+  }
+
+  @Test
+  public void retriedUpdatingListenerJobShouldHaveCorrectHeadersFromUpdateCommand() {
+    testRetriedListenerJobHasExpectedHeaders(
+        ZeebeTaskListenerEventType.updating,
+        pik ->
+            ENGINE
+                .userTask()
+                .ofInstance(pik)
+                .withCandidateUsers("bob", "alice")
+                .withDueDate("new_due_date")
+                .withPriority(88)
+                .withAction("custom_update")
+                .update(),
+        Map.of(
+            Protocol.USER_TASK_CANDIDATE_USERS_HEADER_NAME,
+            "[\"bob\",\"alice\"]",
+            Protocol.USER_TASK_DUE_DATE_HEADER_NAME,
+            "new_due_date",
+            Protocol.USER_TASK_PRIORITY_HEADER_NAME,
+            "88",
+            Protocol.USER_TASK_CHANGED_ATTRIBUTES_HEADER_NAME,
+            "[\"candidateUsersList\",\"dueDate\",\"priority\"]",
+            Protocol.USER_TASK_ACTION_HEADER_NAME,
+            "custom_update"));
+  }
+
+  /**
+   * Verifies that retried listener job has access to the original command's changed attributes by
+   * asserting the presence of expected custom headers on retried, and subsequent listener job.
+   */
+  private void testRetriedListenerJobHasExpectedHeaders(
+      final ZeebeTaskListenerEventType eventType,
+      final Consumer<Long> transitionTrigger,
+      final Map<String, String> expectedHeaders) {
+
+    // given: process with two listener jobs of the same type (first with a failing expression)
+    final long processInstanceKey =
+        helper.createProcessInstance(
+            helper.createProcessWithZeebeUserTask(
+                task ->
+                    task.zeebeTaskListener(l -> l.eventType(eventType).type(listenerType))
+                        .zeebeTaskListener(
+                            l -> l.eventType(eventType).typeExpression("missing_var_name"))));
+
+    // when: trigger transition that causes listener jobs to be created
+    transitionTrigger.accept(processInstanceKey);
+
+    // then: assert first listener job uses expected headers and complete it
+    final Consumer<JobRecordValue> assertHeaders =
+        job ->
+            assertThat(job.getCustomHeaders())
+                .describedAs(
+                    "Listener job should contain headers changed by the original user task command")
+                .containsAllEntriesOf(expectedHeaders);
+
+    helper.assertActivatedJob(processInstanceKey, listenerType, assertHeaders);
+    ENGINE.job().ofInstance(processInstanceKey).withType(listenerType).complete();
+
+    // then: assert incident due to missing variable in listener property expression
+    final var incident =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(incident.getValue())
+        .hasProcessInstanceKey(processInstanceKey)
+        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
+        .hasErrorMessage(
+            """
+                Expected result of the expression 'missing_var_name' to be 'STRING', but was 'NULL'. \
+                The evaluation reported the following warnings:
+                [NO_VARIABLE_FOUND] No variable found with name 'missing_var_name'""");
+
+    // when: fix missing variable and resolve the incident
+    ENGINE
+        .variables()
+        .ofScope(processInstanceKey)
+        .withDocument(Map.of("missing_var_name", listenerType + "_2"))
+        .update();
+    ENGINE.incident().ofInstance(processInstanceKey).withKey(incident.getKey()).resolve();
+
+    // then: retried and remaining listener jobs should receive the same headers
+    helper.assertActivatedJob(processInstanceKey, listenerType, assertHeaders);
+    helper.completeRecreatedJobWithType(processInstanceKey, listenerType);
+    helper.assertActivatedJob(processInstanceKey, listenerType + "_2", assertHeaders);
   }
 }

